File: erlang/bfs.erl
-module(bfs).

-export([visit/2]).

visit(Start, Nodes) ->
    visit(Start, Nodes, []).

visit(Node, Nodes, Visited) ->
    visit_frontier([Node], Nodes, Visited).

visit_frontier([]=_, _, Visited) ->
    Visited;

visit_frontier([_|_]=Frontier, Nodes, _Visited) ->
    Visited = _Visited ++ Frontier,
    Fn = fun(N) -> not(lists:member(N, Visited)) end,
    Next = lists:filter(Fn, frontier_adj(Frontier, Nodes)),
    visit_frontier(Next, Nodes, Visited).

frontier_adj(Frontier, Nodes) ->
    Fn = fun(N, Acc) -> Acc ++ proplists:get_value(N, Nodes) end,
    lists:foldl(Fn, [], Frontier).

File: julia/bfs.jl
module Bfs

    export visit

    function visit(start::Number, nodes::Dict{Int,Array{Int,1}}, fn::Function)
        frontier = Number[start]
        visited  = Number[]

        function next_visit(node::Number)
            filter(n -> !in(n, visited), nodes[node])
        end

        while 0 < length(frontier)
            next = Number[]
            for node = frontier
                push!(visited, node)
                append!(next, next_visit(node))
                fn(node)
            end
            frontier = next
        end
    end

end
